File: backend/utils/filename_parser.py
"""Filename → meeting metadata parser.

Pure-Python, no I/O. Used at upload time to pre-fill meeting_date /
meeting_time / title on new sessions, and exposed as an endpoint so the
frontend can preview the parsed values before submitting.

Patterns, in priority order:

1. Mac Notes + Voice Memos export:
       {notes|downloads}__YYYY-MM-DD_HHMMSS__{title}.{ext}
   Confidence 1.0. All 528 files in Aaron's 2024-09 → 2026-05 audio
   archive match this shape.

2. ISO date prefix with dash separator:
       YYYY-MM-DD - Title.ext
   Confidence 0.9.

3. ISO date prefix, loose separator (underscore or space):
       YYYY-MM-DD_Title.ext
       YYYY-MM-DD Title.ext
   Confidence 0.8.

4. US date suffix with M-D-YY or M-D-YYYY:
       Title - 3-15-24.ext
       Title 3-15-2024.ext
   Confidence 0.7.

5. ISO date anywhere in the basename, no positional contract:
       Anything-with-2024-03-15-buried-inside.ext
   Confidence 0.5.

No match → confidence 0.0, title = best-effort stem cleanup.

The parser is the single source of truth for both the upload pipeline
and the /api/recordings/parse-filename endpoint. The TypeScript twin in
frontend/src/utils/filenameParser.ts mirrors the same five patterns so
the client preview matches what the server stores.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import date, time
from pathlib import Path
from typing import Optional


@dataclass
class ParsedFilename:
    """Result of parse_filename. Confidence reflects which pattern matched."""

    title: Optional[str]
    meeting_date: Optional[date]
    meeting_time: Optional[time]
    source: Optional[str]
    raw_filename: str
    confidence: float


# Pattern 1: Mac Notes + Voice Memos export.
# {notes|downloads}__YYYY-MM-DD_HHMMSS__{title}.{ext}
_PATTERN_NOTES_EXPORT = re.compile(
    r"^(?P<source>notes|downloads)"
    r"__(?P<date>\d{4}-\d{2}-\d{2})"
    r"_(?P<time>\d{6})"
    r"__(?P<title>.+?)"
    r"\.(?P<ext>[A-Za-z0-9]{1,8})$"
)

# Pattern 2: ISO prefix + " - " separator.
_PATTERN_ISO_PREFIX_DASH = re.compile(
    r"^(?P<date>\d{4}-\d{2}-\d{2})\s*-\s*(?P<title>.+?)$"
)

# Pattern 3: ISO prefix with underscore or single space.
_PATTERN_ISO_PREFIX_LOOSE = re.compile(
    r"^(?P<date>\d{4}-\d{2}-\d{2})[ _](?P<title>.+?)$"
)

# Pattern 4: US date suffix. Accepts M-D-YY / MM-DD-YY / M-D-YYYY etc.
# Title sits before the date, with optional " - " or whitespace.
_PATTERN_US_DATE_SUFFIX = re.compile(
    r"^(?P<title>.+?)"
    r"(?:\s*[-_]\s*|\s+)"
    r"(?P<month>\d{1,2})[-/](?P<day>\d{1,2})[-/](?P<year>\d{2}|\d{4})"
    r"$"
)

# Pattern 5: ISO date anywhere in the basename.
_PATTERN_ISO_ANYWHERE = re.compile(r"(\d{4})-(\d{2})-(\d{2})")


def _safe_date(year: int, month: int, day: int) -> Optional[date]:
    try:
        return date(year, month, day)
    except (TypeError, ValueError):
        return None


def _safe_time(hh: int, mm: int, ss: int) -> Optional[time]:
    try:
        return time(hh, mm, ss)
    except (TypeError, ValueError):
        return None


def _normalize_title(raw: str) -> str:
    """Tidy a raw title fragment for display.

    - collapse runs of whitespace / underscores
    - strip leading/trailing punctuation we never want in a title
    - cap at 200 chars (matches the DB column ceiling)
    """
    if not raw:
        return ""
    cleaned = raw.replace("_", " ")
    cleaned = re.sub(r"\s+", " ", cleaned).strip(" -._")
    return cleaned[:200]


def _expand_two_digit_year(year_raw: str) -> int:
    """Two-digit year → four-digit. ≥70 maps to 19YY, <70 to 20YY.

    A pragmatic split that handles meeting filenames going back into the
    late 90s without mis-bucketing anything from this decade.
    """
    n = int(year_raw)
    if len(year_raw) == 2:
        return 1900 + n if n >= 70 else 2000 + n
    return n


def parse_filename(filename: str) -> ParsedFilename:
    """Parse a filename into title / meeting_date / meeting_time / source.

    Always returns a ParsedFilename — never raises. confidence=0.0 means
    no pattern matched and the caller should treat all fields as
    untrusted.
    """
    raw = filename or ""
    # Strip directory, in case a caller passes a path.
    basename = Path(raw).name if raw else ""
    # Pattern 1 is the only one that owns the full filename including
    # the extension (the {ext} group is part of the regex). Try it first.
    m = _PATTERN_NOTES_EXPORT.match(basename)
    if m:
        d = _safe_date(
            int(m.group("date")[0:4]),
            int(m.group("date")[5:7]),
            int(m.group("date")[8:10]),
        )
        t_raw = m.group("time")
        t = _safe_time(int(t_raw[0:2]), int(t_raw[2:4]), int(t_raw[4:6]))
        return ParsedFilename(
            title=_normalize_title(m.group("title")),
            meeting_date=d,
            meeting_time=t,
            source=m.group("source"),
            raw_filename=raw,
            confidence=1.0,
        )

    # Patterns 2-5 work on the stem (no extension). Strip the last
    # extension only — preserving any dots in the title itself.
    stem = Path(basename).stem if basename else ""

    m = _PATTERN_ISO_PREFIX_DASH.match(stem)
    if m:
        d = _safe_date(
            int(m.group("date")[0:4]),
            int(m.group("date")[5:7]),
            int(m.group("date")[8:10]),
        )
        if d:
            return ParsedFilename(
                title=_normalize_title(m.group("title")),
                meeting_date=d,
                meeting_time=None,
                source="generic",
                raw_filename=raw,
                confidence=0.9,
            )

    m = _PATTERN_ISO_PREFIX_LOOSE.match(stem)
    if m:
        d = _safe_date(
            int(m.group("date")[0:4]),
            int(m.group("date")[5:7]),
            int(m.group("date")[8:10]),
        )
        if d:
            return ParsedFilename(
                title=_normalize_title(m.group("title")),
                meeting_date=d,
                meeting_time=None,
                source="generic",
                raw_filename=raw,
                confidence=0.8,
            )

    m = _PATTERN_US_DATE_SUFFIX.match(stem)
    if m:
        d = _safe_date(
            _expand_two_digit_year(m.group("year")),
            int(m.group("month")),
            int(m.group("day")),
        )
        if d:
            return ParsedFilename(
                title=_normalize_title(m.group("title")),
                meeting_date=d,
                meeting_time=None,
                source="generic",
                raw_filename=raw,
                confidence=0.7,
            )

    m = _PATTERN_ISO_ANYWHERE.search(stem)
    if m:
        d = _safe_date(int(m.group(1)), int(m.group(2)), int(m.group(3)))
        if d:
            # Strip the matched date out of the stem to recover a title.
            without_date = (stem[: m.start()] + stem[m.end():]).strip(" -_.")
            title = _normalize_title(without_date) or _normalize_title(stem)
            return ParsedFilename(
                title=title,
                meeting_date=d,
                meeting_time=None,
                source="generic",
                raw_filename=raw,
                confidence=0.5,
            )

    # No pattern matched. Return a best-effort title (cleaned stem) and
    # let the caller decide whether to fall back to mtime or a default.
    fallback_title = _normalize_title(stem) or None
    return ParsedFilename(
        title=fallback_title,
        meeting_date=None,
        meeting_time=None,
        source=None,
        raw_filename=raw,
        confidence=0.0,
    )


# Regex for the "Call with {Name}" pattern used by speaker auto-link (B-import.3).
_CALL_WITH_RE = re.compile(r"^\s*Call\s+with\s+(.+?)\s*$", re.IGNORECASE)


def extract_call_with_name(title: Optional[str]) -> Optional[str]:
    """If *title* matches ``Call with {Name}`` (case-insensitive), return the
    cleaned name. Handles parenthetical suffixes like ``Doug (Crash)`` by
    stripping the parenthetical for matching purposes, returning ``Doug``.
    Returns ``None`` for any other format or a falsy input.
    """
    if not title:
        return None
    m = _CALL_WITH_RE.match(title)
    if not m:
        return None
    name = m.group(1).strip()
    # Strip a trailing parenthetical like " (Crash)" so 'Doug (Crash)'
    # matches an enrolled speaker whose display_name is 'Doug'.
    name = re.sub(r"\s*\([^)]*\)\s*$", "", name).strip()
    return name or None


if __name__ == "__main__":
    samples = [
        # Pattern 1 — real-world Mac Notes + Voice Memos export.
        "notes__2024-11-05_233322__Call with Jason Allen.m4a",
        "notes__2024-11-14_183247__Call with John Rahaghi.m4a",
        "downloads__2026-05-16_212900__Call with Shafen Khan.m4a",
        "downloads__2026-05-14_161900__Legacy25 Capital.m4a",
        # Pattern 1 — title with parentheses.
        "notes__2025-03-04_091500__Doug (Crash).m4a",
        # Pattern 2 — ISO prefix with " - ".
        "2024-03-15 - Jane Smith.txt",
        # Pattern 3 — ISO prefix loose.
        "2024-03-15_Jane Smith.txt",
        "2024-03-15 Jane Smith Q1 sync.m4a",
        # Pattern 4 — US date suffix.
        "Jane Smith - 3-15-24.txt",
        "Quarterly review 3-15-2024.txt",
        # Pattern 5 — ISO date buried.
        "zoom_2024-03-15_recording.mp4",
        # No match.
        "Just a random title.m4a",
        "",
    ]
    for f in samples:
        p = parse_filename(f)
        print(
            f"{p.confidence:.1f}  date={p.meeting_date}  time={p.meeting_time}  "
            f"src={p.source!s:>10}  title={p.title!r}  file={f!r}"
        )

File: frontend/src/utils/filenameParser.ts
/**
 * Filename → meeting metadata parser (TypeScript twin).
 *
 * Mirrors `backend/utils/filename_parser.py` exactly. Used at single-file
 * upload time so the new-session form can show a preview ("This looks like
 * a meeting from 2024-11-05 with Jason Allen — use these values?") before
 * the user commits. The backend re-parses on its own at upload finalize
 * and uses the user's overrides where provided, so a frontend hiccup
 * here never poisons stored metadata.
 *
 * Patterns, in priority order:
 *   1. Mac Notes + Voice Memos export    confidence 1.0
 *      {notes|downloads}__YYYY-MM-DD_HHMMSS__{title}.{ext}
 *   2. ISO date prefix with " - "        confidence 0.9
 *      YYYY-MM-DD - Title.ext
 *   3. ISO prefix loose separator        confidence 0.8
 *      YYYY-MM-DD_Title.ext / YYYY-MM-DD Title.ext
 *   4. US date suffix M-D-YY[YY]         confidence 0.7
 *      Title - M-D-YY.ext / Title M-D-YYYY.ext
 *   5. ISO date buried in basename       confidence 0.5
 *   no match → confidence 0.0
 */

export type FilenameSource = "notes" | "downloads" | "generic" | null;

export interface ParsedFilename {
  /** Cleaned-up display title, or null if the basename was empty. */
  title: string | null;
  /** ISO 8601 YYYY-MM-DD when a valid calendar date was extracted. */
  meetingDate: string | null;
  /** 24h HH:MM:SS when the filename carried a time component. */
  meetingTime: string | null;
  /** Which pattern's source label this came from. */
  source: FilenameSource;
  /** 0.0–1.0 — which pattern matched. */
  confidence: number;
  /** Echo of the original input for traceability. */
  rawFilename: string;
}

const PATTERN_NOTES_EXPORT =
  /^(?<source>notes|downloads)__(?<date>\d{4}-\d{2}-\d{2})_(?<time>\d{6})__(?<title>.+?)\.(?<ext>[A-Za-z0-9]{1,8})$/;
const PATTERN_ISO_PREFIX_DASH = /^(?<date>\d{4}-\d{2}-\d{2})\s*-\s*(?<title>.+?)$/;
const PATTERN_ISO_PREFIX_LOOSE = /^(?<date>\d{4}-\d{2}-\d{2})[ _](?<title>.+?)$/;
const PATTERN_US_DATE_SUFFIX =
  /^(?<title>.+?)(?:\s*[-_]\s*|\s+)(?<month>\d{1,2})[-/](?<day>\d{1,2})[-/](?<year>\d{2}|\d{4})$/;
const PATTERN_ISO_ANYWHERE = /(\d{4})-(\d{2})-(\d{2})/;

/** Strip directory prefixes — works for forward and back slashes. */
function basenameOf(input: string): string {
  if (!input) return "";
  const lastSlash = Math.max(input.lastIndexOf("/"), input.lastIndexOf("\\"));
  return lastSlash >= 0 ? input.slice(lastSlash + 1) : input;
}

/** Strip the last `.ext` from a filename (preserves dots inside the name). */
function stemOf(name: string): string {
  if (!name) return "";
  const dot = name.lastIndexOf(".");
  return dot > 0 ? name.slice(0, dot) : name;
}

function isValidDate(y: number, m: number, d: number): boolean {
  if (m < 1 || m > 12 || d < 1 || d > 31) return false;
  const dt = new Date(Date.UTC(y, m - 1, d));
  return (
    dt.getUTCFullYear() === y &&
    dt.getUTCMonth() === m - 1 &&
    dt.getUTCDate() === d
  );
}

function isValidTime(h: number, m: number, s: number): boolean {
  return h >= 0 && h < 24 && m >= 0 && m < 60 && s >= 0 && s < 60;
}

function pad2(n: number): string {
  return n < 10 ? `0${n}` : String(n);
}

function isoDate(y: number, m: number, d: number): string {
  return `${y}-${pad2(m)}-${pad2(d)}`;
}

function isoTime(h: number, m: number, s: number): string {
  return `${pad2(h)}:${pad2(m)}:${pad2(s)}`;
}

function normalizeTitle(raw: string): string {
  if (!raw) return "";
  let cleaned = raw.replace(/_/g, " ");
  cleaned = cleaned.replace(/\s+/g, " ").trim();
  // Strip leading/trailing punctuation we never want in a meeting title.
  cleaned = cleaned.replace(/^[\s\-._]+|[\s\-._]+$/g, "");
  return cleaned.slice(0, 200);
}

function expandTwoDigitYear(yearRaw: string): number {
  const n = parseInt(yearRaw, 10);
  if (yearRaw.length === 2) {
    return n >= 70 ? 1900 + n : 2000 + n;
  }
  return n;
}

/**
 * Parse `filename` into title / meeting date / meeting time / source.
 *
 * Never throws. confidence = 0.0 means no pattern matched and the
 * caller should treat `meetingDate` / `meetingTime` as unknown.
 */
export function parseFilename(filename: string): ParsedFilename {
  const raw = filename || "";
  const basename = basenameOf(raw);

  // Pattern 1 — owns the .ext group so it's checked on the full basename.
  const m1 = basename.match(PATTERN_NOTES_EXPORT);
  if (m1 && m1.groups) {
    const dateStr = m1.groups.date;
    const timeStr = m1.groups.time;
    const y = parseInt(dateStr.slice(0, 4), 10);
    const mo = parseInt(dateStr.slice(5, 7), 10);
    const d = parseInt(dateStr.slice(8, 10), 10);
    const hh = parseInt(timeStr.slice(0, 2), 10);
    const mm = parseInt(timeStr.slice(2, 4), 10);
    const ss = parseInt(timeStr.slice(4, 6), 10);

    return {
      title: normalizeTitle(m1.groups.title) || null,
      meetingDate: isValidDate(y, mo, d) ? isoDate(y, mo, d) : null,
      meetingTime: isValidTime(hh, mm, ss) ? isoTime(hh, mm, ss) : null,
      source: m1.groups.source as "notes" | "downloads",
      confidence: 1.0,
      rawFilename: raw,
    };
  }

  // Patterns 2-5 operate on the stem (no extension).
  const stem = stemOf(basename);

  const m2 = stem.match(PATTERN_ISO_PREFIX_DASH);
  if (m2 && m2.groups) {
    const y = parseInt(m2.groups.date.slice(0, 4), 10);
    const mo = parseInt(m2.groups.date.slice(5, 7), 10);
    const d = parseInt(m2.groups.date.slice(8, 10), 10);
    if (isValidDate(y, mo, d)) {
      return {
        title: normalizeTitle(m2.groups.title) || null,
        meetingDate: isoDate(y, mo, d),
        meetingTime: null,
        source: "generic",
        confidence: 0.9,
        rawFilename: raw,
      };
    }
  }

  const m3 = stem.match(PATTERN_ISO_PREFIX_LOOSE);
  if (m3 && m3.groups) {
    const y = parseInt(m3.groups.date.slice(0, 4), 10);
    const mo = parseInt(m3.groups.date.slice(5, 7), 10);
    const d = parseInt(m3.groups.date.slice(8, 10), 10);
    if (isValidDate(y, mo, d)) {
      return {
        title: normalizeTitle(m3.groups.title) || null,
        meetingDate: isoDate(y, mo, d),
        meetingTime: null,
        source: "generic",
        confidence: 0.8,
        rawFilename: raw,
      };
    }
  }

  const m4 = stem.match(PATTERN_US_DATE_SUFFIX);
  if (m4 && m4.groups) {
    const y = expandTwoDigitYear(m4.groups.year);
    const mo = parseInt(m4.groups.month, 10);
    const d = parseInt(m4.groups.day, 10);
    if (isValidDate(y, mo, d)) {
      return {
        title: normalizeTitle(m4.groups.title) || null,
        meetingDate: isoDate(y, mo, d),
        meetingTime: null,
        source: "generic",
        confidence: 0.7,
        rawFilename: raw,
      };
    }
  }

  const m5 = stem.match(PATTERN_ISO_ANYWHERE);
  if (m5) {
    const y = parseInt(m5[1], 10);
    const mo = parseInt(m5[2], 10);
    const d = parseInt(m5[3], 10);
    if (isValidDate(y, mo, d)) {
      // Recover a title by removing the matched date from the stem.
      const before = stem.slice(0, m5.index!);
      const after = stem.slice(m5.index! + m5[0].length);
      const titleRaw = (before + " " + after).trim();
      const title = normalizeTitle(titleRaw) || normalizeTitle(stem);
      return {
        title: title || null,
        meetingDate: isoDate(y, mo, d),
        meetingTime: null,
        source: "generic",
        confidence: 0.5,
        rawFilename: raw,
      };
    }
  }

  // No pattern matched — best-effort title from the cleaned stem.
  const fallback = normalizeTitle(stem);
  return {
    title: fallback || null,
    meetingDate: null,
    meetingTime: null,
    source: null,
    confidence: 0.0,
    rawFilename: raw,
  };
}
